### DISCUSSION TITLE
Plot creation times on Apple M1

### DISCUSSION THREAD

**Discussion by AndyRPH** [madMAx43v3r/chia-plotter#169] [General]
Its come up in a few issues, so figured I'd make a discussion spot for plot creation times on Apple M1 CPUs. 

The latest build is now giving me times around 17k seconds, so not great, actually, slower than the pechy-optimised plotter I was previously using.  This is running it for 4 threads using tmp1 on a usb 3.1 NVME SSD and tmp2 on a TB3 NVME SSD.  The MBPro has 16 GB ram and not much else going on during these times. 

Two days ago, I was getting consistent plots of 3 hours or so, so it's slowed down a bit. Not sure if it was the blake3 commit or something else. 

For the record, I even did one plot using my internal SSD and that too didn't get much faster than 3 hours for a plot.  Plots using 2 threads are a bit slower than 4 threads but not by much.  

Seems the M1 CPU machines are bottlenecked elsewhere than this plotter will help with.

**Comment by stingray072** [comment]
> Its come up in a few issues, so figured I'd make a discussion spot for plot creation times on Apple M1 CPUs.
> 
> The latest build is now giving me times around 17k seconds, so not great, actually, slower than the pechy-optimised plotter I was previously using. This is running it for 4 threads using tmp1 on a usb 3.1 NVME SSD and tmp2 on a TB3 NVME SSD. The MBPro has 16 GB ram and not much else going on during these times.
> 
> Two days ago, I was getting consistent plots of 3 hours or so, so it's slowed down a bit. Not sure if it was the blake3 commit or something else.
> 
> For the record, I even did one plot using my internal SSD and that too didn't get much faster than 3 hours for a plot. Plots using 2 threads are a bit slower than 4 threads but not by much.
> 
> Seems the M1 CPU machines are bottlenecked elsewhere than this plotter will help with.

Try setting threads to 200 and see what happens. Seemed to give some good results in harnessing the M1 on prior plotters.

**Reply by AndyRPH** [reply]
It's a four core processor...(plus four "Efficency" cores). Why would you set the threads to 200?

**Reply by dmorrison91** [reply]
Before Max I also saw a person do 200 threads, and so I did and it sped up my plots incredibly and only used 30% of the CPU roughly. I don't know maybe arm process timer runs very different, also saw that you can go up to 2500 threads in the activity monitor so until you hit that limit... run multi-instances.

**Reply by stingray072** [reply]
Yup, something to do with the way M1 translates things. Don’t have to try if you don’t want to, just a suggestion. Would be curious what the time comes down to

**Reply by AndyRPH** [reply]
Okay, for fun, I ran it with it set to 16 threads.  Slowed to a crawl.  After 3 hours it hadn't even finished phase 1 and I killed it.

**Comment by dmorrison91** [comment]
Also how'd you compile, I'm getting an error message on the compile on my m1 mini.

**Reply by AndyRPH** [reply]
What error, try again with latest commits.

**Reply by dmorrison91** [reply]
I'm getting a lot of unexpected arguments with blake3_avx512_x86-64_unix.s

**Reply by AndyRPH** [reply]
With the latest commits?  I thought they fixed that last night.

**Reply by dmorrison91** [reply]
Yea I deleted everything and restarted the entire compile, but still getting this error now

`/Users/derekmorrison/chia-plotter/src/chia_plot.cpp:16:10: fatal error: 'sodium.h' file not found
#include <sodium.h>
         ^~~~~~~~~~
1 error generated.
make[2]: *** [CMakeFiles/chia_plot.dir/src/chia_plot.cpp.o] Error 1
make[1]: *** [CMakeFiles/chia_plot.dir/all] Error 2
make: *** [all] Error 2
`

**Reply by dmorrison91** [reply]
I've tried brew install libsodium-dev but get a no similarly named formula error

**Reply by leecarney** [reply]
On m1 I've found the libraries are stored in /opt/homebrew/Cellar/libsodium/1.0.18_1/include so you need to use these in your ln -s commands.
That being said I still have an ld -lsodium error (used on the last line of Cmakefile) that I just can't get past no matter what LDFLAGS options etc I use.

**Reply by dmorrison91** [reply]
Okay made the changes now I'm getting new errors:

`[100%] Linking CXX executable chia_plot
clang: warning: argument unused during compilation: '-fmax-errors=1' [-Wunused-command-line-argument]
ld: library not found for -lsodium
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [chia_plot] Error 1
make[1]: *** [CMakeFiles/chia_plot.dir/all] Error 2
make: *** [all] Error 2`

**Reply by moertel187** [reply]
have also these errors

**Comment by xxxxxxPETExxxxx** [comment]
Just as comparison I'm using Plotman with regular chia plotter on M1 MBP with 4 parallel and getting 18-19 plots per day on a 1TB Samsung 980 pro in TB3. So If you're getting 3 hours using MadMax I think you should switch to Plotman since my average time per plot is much lower than yours.

**Reply by AndyRPH** [reply]
Care to share your plotman settings?  I hadn't really explored using plotman on M1, I was doing it via the GUI, but compiled the branch from here: https://github.com/pechy/chiapos/tree/combined

**Reply by xxxxxxPETExxxxx** [reply]
Sure: copy and paste of my plotman:

   tmpdir_stagger_phase_major: 2
        tmpdir_stagger_phase_minor: 0
        # Optional: default is 1
        tmpdir_stagger_phase_limit: 2

        # Don't run more than this many jobs at a time on a single temp dir.
        tmpdir_max_jobs: 4

        # Don't run more than this many jobs at a time in total.
        global_max_jobs: 4

        # Don't run any jobs (across all temp dirs) more often than this, in minutes.
        global_stagger_m: 75

        # How often the daemon wakes to consider starting a new plot job, in seconds.
        polling_time_s: 120

        k: 32
        e: False             # Use -e plotting option
        n_threads: 8         # Threads per job
        n_buckets: 128       # Number of buckets to split data into
        job_buffer: 3800     # Per job memory

**Reply by xxxxxxPETExxxxx** [reply]
I'm experimenting on my AMD r9 5900x at staggering 3 or 4 plots on madmax to see if I can get a lower average time on that, but it uses quite a lot of ram so I don't think it's possible to run more than one at a time in madmax on my 16GB M1

**Reply by AndyRPH** [reply]
You don't have issues giving it 8 threads when there are only 4 performance cores on the M1?  Are you using plotman with the official chia binaries, or one of the optimized ones?

**Reply by dmorrison91** [reply]
Andy, here's my use of it, I have my m1 I was using 8 threads to begin with using plotman, I happened upon a thread and the guy put in 200 threads. I put in 200 threads, I went from 16-20hrs down to 8-9hrs. If you put in 200 threads and it doesn't work just cancel it.

**Reply by xxxxxxPETExxxxx** [reply]
no problems with 8 threads each. If you look in Activity monitor your Mac has more than 1500 threads going at any given time. I'm using plotman with the official plotter. My single plot time is just over 3 hours. with 4 parallel it's 4.5 hours. Can't imagine what your bottleneck is for 8-9 hours or 20 hours. Must be really slow temp drive.

**Reply by AndyRPH** [reply]
Oh on the pechy optimized plotter I can do 3hr/plot with 4 threads, if I run two staggered I can get it to 4.5 hr/s per plot x2.   Interesting on the threads, it did NOT work at all that way with max's, plotter, 16 threads dropped speed dramatically, but I'll try that in plotman and see what happens.

**Reply by xxxxxxPETExxxxx** [reply]
Post back your results please. I had never heard of this optimized plotter.

**Reply by AndyRPH** [reply]
I linked to it above, it's about 30% faster than the official plotter, but works differently than max's plotter I believe.  We'll see tomorrow if giving plotman using that optimized plotter 16 threads makes a difference over 2/4 threads.

**Reply by xxxxxxPETExxxxx** [reply]
Thanks. Another program I use is "TG Pro" to monitor CPU temps and "auto boost" the fans. you can try it free for two weeks and after that it's only $12. Keeps my temps much lower than without which I assume helps to avoid CPU throttling.
